src/miner.cpp

```diff
@@ -230,16 +231,21 @@ void BlockAssembler::AddToBlock(CTxMemPool::txiter iter)
     }
 }
 
-int BlockAssembler::UpdatePackagesForAdded(const CTxMemPool::setEntries& alreadyAdded,
+template<typename SortedIterable>
+int BlockAssembler::UpdatePackagesForAdded(const SortedIterable& alreadyAdded,
         indexed_modified_transaction_set &mapModifiedTx)
 {
     int nDescendantsUpdated = 0;
+    CTxMemPool::vecEntries descendants;
     for (CTxMemPool::txiter it : alreadyAdded) {
-        CTxMemPool::setEntries descendants;
-        mempool.CalculateDescendants(it, descendants);
+        descendants.clear();
+        // can't use external epoch to loop because we want to update
+        // all descendants
+        // No need to add self (it) because we would filter it from the loop
+        mempool.CalculateDescendantsVec(it, descendants);
         // Insert all descendants (not yet in block) into the modified set
         for (CTxMemPool::txiter desc : descendants) {
-            if (alreadyAdded.count(desc))
+            if (std::binary_search(alreadyAdded.cbegin(), alreadyAdded.cend(), desc, CTxMemPool::CompareIteratorByHash()))
```

Comment:
std::binary_search on a std::set takes `O(N log N)` because std::advance on the iterators is `O(N)`.

To fix this, allow passing in a predicate that knows how to query alreadyAdded efficient. Other options would be to make alreadyAdded a sorted vector or make it a hashtable always. But that's a bigger change than the predicate, which doesn't change any data types.